Add `freeze!`/`thaw!` (#112)
FluxML/Optimisers.jl: docs/src/index.md

st = Optimisers.setup(DecayDescent(0.1), Layer(3))
```

## Frozen Parameters

To temporarily prevent training from affecting some parameters,
use [freeze!](@ref Optimisers.freeze!) and `thaw!`.

> They work by mutating all `Leaf`s of the state tree, within the given limb:

Reviewer: ```suggestion
They work by mutating all `Leaf`s of the state tree, within the (sub-)tree passed in:
```

Author: You think people won't be thinking of 🌳 by the end of the sentence? I was going to say "branch" but that has another meaning... maybe I'd rather avoid naming things:
```suggestion
They work by mutating all `Leaf`s of the state tree, or part of it:
```

Reviewer: Yeah your suggestion is best I think

Author: Done.

```julia
using Flux, Optimisers

x = randn(Float32, 28, 28, 1, 1);
net = @autosize (size(x)...,) Chain(
  Conv((3, 3), 1 => 3, stride=2, bias=false), Flux.flatten, Dense(_ => 2, relu),
)
opt = Optimisers.setup(Optimisers.Momentum(), net);

net.layers[3] isa Dense  # now freeze this layer's parameters:

> Optimisers.freeze!(opt.layers[3])

Reviewer: ```suggestion
Optimisers.freeze!(opt.layers[3]) # pass a sub-tree in for freezing
```

Author: I think this is addressed by the line above? The present state avoids giving a name to what the part of a tree is.

Optimisers.update!(opt, net, gradient(m -> sum(m(x)), net)...);

> opt  # bias = Leaf(Momentum{Float32}(0.01, 0.9), Float32[0.0, 0.0], frozen = true)

Reviewer 2: This would be nice as a doctest.

Author: I tried & now everything is broken locally, I don't understand Documenter. 

Note that the tests at present don't load Flux at all. And that I think testing the output of this block probably means testing a bunch of Flux printing which I'm not sure this package should ever break if that changes.

Reviewer 2: Fair point. In that case, this line could be eliminated entirely and the comment moved to directly before/after the `freeze!` call. Also duplicating it after the `thaw!` call to show `frozen = false`. 
```julia
opt_state = Optimisers.setup(Optimisers.Momentum(), net);

...

Optimisers.freeze!(opt_state.layers[3])
# opt_state.layers[3].bias = Leaf(Momentum{Float32}(0.01, 0.9), Float32[0.0, 0.0], frozen = true)

...

Optimisers.thaw!(opt)
# opt_state.layers[3].bias = Leaf(Momentum{Float32}(0.01, 0.9), Float32[0.0, 0.0], frozen = false)
...
```

Author: Last commit does something like this

st = Optimisers.setup(DecayDescent(0.1), Layer(3))
```

## Frozen Parameters

To temporarily prevent training from affecting some parameters,
use [freeze!](@ref Optimisers.freeze!) and `thaw!`.
They work by mutating all `Leaf`s of the state tree, or part of it.

```julia
using Flux, Optimisers

x = randn(Float32, 28, 28, 1, 1);

> net = @autosize (size(x)...,) Chain(

Reviewer 3: what's the point of splatting here instead of just using `size(x)`?

Author: The macro at present wants to see a tuple expression, not a call. It could be made to accept anything.
